gui/src/app/RunPanel/RunPanel.tsx

```diff
@@ -1,153 +1,113 @@
 /* eslint-disable @typescript-eslint/no-explicit-any */
-import { FunctionComponent, useCallback, useEffect, useState } from 'react';
+import Box from '@mui/material/Box';
 import LinearProgress, { LinearProgressProps } from '@mui/material/LinearProgress';
 import Typography from '@mui/material/Typography';
-import Box from '@mui/material/Box';
+import { FunctionComponent, useCallback, useEffect, useState } from 'react';
 
-import StanWorker from '../tinystan/Worker?worker';
-import { Progress, Replies, Requests } from '../tinystan/Worker'
+import StanSampler, { StanSamplerStatus } from '../StanSampler/StanSampler';
+import { Progress } from '../tinystan/Worker';
 
 type RunPanelProps = {
     width: number;
     height: number;
-    compiledUrl: string;
+    sampler?: StanSampler;
     data: any | undefined
     dataIsSaved: boolean
 };
 
-type RunStatus = '' | 'waiting' | 'running' | 'done' | 'failed';
+const numChains = 4;
 
-const chains = 4;
+const RunPanel: FunctionComponent<RunPanelProps> = ({ width, height, sampler, data, dataIsSaved }) => {
 
-const RunPanel: FunctionComponent<RunPanelProps> = ({ width, height, compiledUrl, data, dataIsSaved }) => {
-
-    const [runStatus, setRunStatus] = useState<RunStatus>('waiting');
+    const [runStatus, setRunStatus] = useState<StanSamplerStatus>('');
     const [errorMessage, setErrorMessage] = useState<string>('');
-    const [samples, setSamples] = useState<number[][]>([[]]);
-    const [paramNames, setParamNames] = useState<string[]>([]);
 
     const [progress, setProgress] = useState<Progress | undefined>(undefined);
 
-    const [modelWorker, setModelWorker] = useState<Worker | undefined>(undefined);
-
-    // Cancellation destroys the worker, and therefore
-    // requires the same code to be run as if a new
-    // URL was provided. This state is therefore
-    // used to force the following useEffect hook to rerun.
-    const [trigger, setTrigger] = useState(false);
-
     useEffect(() => {
-        if (!compiledUrl) {
-            setModelWorker(undefined);
-            return;
-        }
-        const worker = new StanWorker();
-        setModelWorker(worker);
-        worker.postMessage({ purpose: Requests.Load, url: compiledUrl });
-        return () => {
-            if (worker) {
-                console.log("Cleaning up worker");
-                worker.terminate();
+        if (!sampler) return;
+        let canceled = false;
+        sampler.onStatusChanged(() => {
+            if (canceled) return;
+            setRunStatus(sampler.status);
+            if (sampler.status === 'failed') {
+                setErrorMessage(sampler.errorMessage);
             }
-        }
-    }, [compiledUrl, trigger])
+        })
+        sampler.onProgress((progress: Progress) => {
+            if (canceled) return;
+            setProgress(progress);
+        })
+        return (
+            () => {
+                canceled = true;
+                sampler.cancel();
```

Comment:
Is the `canceled` variable really necessary? After `terminate` is called I wouldn't expect any further callbacks to fire

Reply:
I think it's safest to keep this in here because you could conceivably have a situation where the sampler prop changes, but the old sampler is still alive and emitting  callbacks (doesn't actually happen in the current program, but we want this RunPanel component to be independent). The `canceled` here doesn't mean that the sampler was canceled, but it means that the useEffect was canceled due to the sampler prop changing.